test/create_game_section/Input_password/input_password_widget_test.dart

```diff
@@ -0,0 +1,114 @@
+      final _mockDatabaseService = MockDatabaseService();
+      final mockGame = Game(
+          id: 'mock_game_123',
+          name: null,
+          creatorName: null,
+          sonarIntervals: null,
+          password: 'test_password',
+          boundary: null,
+          phase: null);
+
+      // init input password page
+      await tester.pumpWidget(MultiProvider(
+        providers: [
+          Provider<AuthService>.value(value: _mockAuthService),
+          Provider<DatabaseService>.value(value: _mockDatabaseService),
+        ],
+        child: MaterialApp(
+          home: InputPasswordScreen(
+            game: mockGame,
+          ),
+        ),
+      ));
+
+      // tap to join a game
+      final passwordFeild = find.byKey(Key('input_password_screen_text_feild'));
+      expect((passwordFeild), findsOneWidget);
+      await tester.enterText(passwordFeild, 'incorrectPassword');
+      await tester.tap(find.byKey(Key('input_password_screen_submit_btn')));
+      await tester.pumpAndSettle();
+
+      // show error message
+      expect(find.text('Password is incorrect'), findsOneWidget);
+    });
+  });
+
+  group('on successfull password submission', () {
+    testWidgets(
+        'add player to the game in firestore then navigate to waiting screen ',
+        (WidgetTester tester) async {
+      final _authController = StreamController<UserData>();
+      final _mockAuthService = MockAuthService(controller: _authController);
+      final _databaseController = StreamController<List<Game>>();
+      final _mockDatabaseService =
+          MockDatabaseService(controller: _databaseController);
+      final mockGame = Game(
+          id: 'mock_game_123',
+          name: null,
+          creatorName: null,
+          sonarIntervals: null,
+          password: 'test_password',
+          boundary: null,
+          phase: null);
+
+      // init input password page
+      await tester.pumpWidget(MultiProvider(
+        providers: [
+          Provider<AuthService>.value(value: _mockAuthService),
+          Provider<DatabaseService>.value(value: _mockDatabaseService),
+        ],
+        child: MaterialApp(
+          home: InputPasswordScreen(
+            game: mockGame,
+          ),
+        ),
+      ));
+
+      // tap to join a game
+      await tester.enterText(
+          find.byKey(Key('input_password_screen_text_feild')), 'test_password');
+      await tester.tap(find.byKey(Key('input_password_screen_submit_btn')));
+      await tester.pumpAndSettle();
+      await tester.pumpAndSettle();
+
+      // check user has been added to game
+      final mockPlayer = Player(
+          id: 'uid123', // given from _mockAuthService.currentUserId
+          username:
+              'username123', // given from _mockDatabaseService.getUsername
+          isTagger: null,
+          hasBeenTagged: null,
+          hasItem: null,
+          isAdmin: null);
+      verify(_mockDatabaseService.joinGame(mockGame.id, mockPlayer))
```

Comment:
here she is, its breaking at this line saying that im using a non mockito object 😢.
Ive tried calling verify(_mockDatabaseService.joinGame(any, any)) and flooding the test file with tester.pumps & tester.pump and settle.

In the input password screen widget, when you tap submit, there are a couple of async functions happening before "joinGame" gets called. im thinking this might be why it cant find it but i really dont know right now haha, been staring at it for a while. think its time for a skate lol